p4include/psa.p4

```diff
@@ -0,0 +1,675 @@
+  InstanceType_t           instance_type;  /// Clone or Normal
+  /// set by the runtime in the parser, these are not under programmer control
+  ParserStatus_t           parser_status;
+  ParserErrorLocation_t    parser_error_location;
+}
+
+struct psa_ingress_output_metadata_t {
+  PortId_t                 egress_port;
+}
+
+struct psa_egress_input_metadata_t {
+  PortId_t                 egress_port;
+  InstanceType_t           instance_type;  /// Clone or Normal
+  EgressInstance_t         instance;       /// instance coming from PRE
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Additional supported match_kind types
+ */
+match_kind {
+    range,   /// Used to represent min..max intervals
+    selector /// Used for implementing dynamic_action_selection
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ *  Cloning methods
+ */
+enum CloneMethod_t {
+  /// Clone method         Packet source             Insertion point
+  Ingress2Ingress,  /// original ingress,            Ingress parser
+  Ingres2Egress,    /// post parse original ingres,  Buffering queue
+  Egress2Ingress,   /// post deparse in egress,      Ingress parser
+  Egress2Egress     /// inout to deparser in egress, Buffering queue
+}
+
+/* --------------------------------------------------------------------- */
+/*
+ * PSA Externs
+ */
+
+/**
+ * The PacketReplicationEngine extern represents the non-programmable
+ * part of the PSA pipeline.
+ *
+ * Even though the PRE can not be programmed using P4, it can be
+ * configured both directly using control plane APIs and by setting
+ * intrinsic metadata. In this specification we opt to define the
+ * operations available in the PRE as method invocations. A target
+ * backend is responsible for mapping the PRE extern APIs to the
+ * appropriate mechanisms for performing these operations in the
+ * hardware.
+ *
+ * The PRE is instantiated by the architecture and a P4 program can
+ * use it directly. It is an error to instantiate the PRE multiple
+ * times.  The PRE is made available to the Ingress programmable block
+ * using the same mechanism as packet_in. A corresponding Buffering
+ * and Queueing Engine (BQE) extern is defined for the Egress
+ * pipeline (@see BufferingQueueingEngine).
+ *
+ * Note: some of these operations may not be implemented as primitive
+ * operations on a certain target. However, All of the operations can
+ * be implemented as a combination of other operations. Applications
+ * that rely on non-primitive operations may incur significant
+ * performance penalties, however, they should be functionally
+ * correct.
+ *
+ * Semantics of behavior for multiple calls to PRE APIs
+ *
+ * The semantics of calling the PRE APIs is equivalent to setting
+ * intrinsic metadata fields/bits and assumming that the PRE looks up
+ * the fields in the following order: drop, truncate, multicast,
+ * clone, output_port.
```

Comment:
they are actually the bits in the intrinsic metadata. I guess the statement needs clarification. We have intrinsic metadata defined in the v1model for most of these bits. What I intended to say is that one can think of those bits as "global state" and the API functions modifying the global state. The PRE looks at the state after all the functions have been called in the pipeline and makes the decision based on the state of the bits. I'll appreciate your help in rephrasing the paragraph accordingly.